File: mac/Clipmon/Clipmon/SyncProtocol.swift
import CryptoKit
import Foundation

/// Wire-format constants and helpers shared with the Windows client.
/// IMPORTANT: keep aligned with `SyncProtocol.cs`.
enum SyncProtocol {
    static let keySalt = "clipmon-key-v1"
    static let roomSalt = "clipmon-room-v1"
    static let pbkdf2Iterations = 200_000
    static let keySizeBytes = 32 // AES-256
    static let nonceBytes = 12
    static let tagBytes = 16

    /// 32-byte AES-256 key derived from the pairing code.
    /// Uses HKDF over SHA-256(pairingCode || salt) instead of PBKDF2 because CommonCrypto's
    /// PBKDF2 is verbose; HKDF is the same cost-class for a single-use key derivation here.
    /// To stay compatible with the Windows side we instead implement PBKDF2-HMAC-SHA256 manually.
    static func deriveKey(pairingCode: String) throws -> SymmetricKey {
        guard !pairingCode.isEmpty else { throw SyncError.missingPairingCode }
        let password = Data(pairingCode.utf8)
        let salt = Data(keySalt.utf8)
        let bytes = pbkdf2HmacSha256(password: password, salt: salt, iterations: pbkdf2Iterations, keyLength: keySizeBytes)
        return SymmetricKey(data: bytes)
    }

    static func deriveRoomId(pairingCode: String) throws -> String {
        guard !pairingCode.isEmpty else { throw SyncError.missingPairingCode }
        let combined = Data((pairingCode + ":" + roomSalt).utf8)
        let digest = SHA256.hash(data: combined)
        return digest.map { String(format: "%02x", $0) }.joined()
    }

    static func encryptEnvelope(key: SymmetricKey, plaintextJson: String) throws -> String {
        let plaintext = Data(plaintextJson.utf8)
        let sealed = try AES.GCM.seal(plaintext, using: key)
        // Wire format: [12 nonce] [ciphertext] [16 tag]  (matches AesGcm in .NET)
        let nonceData: Data = sealed.nonce.withUnsafeBytes { Data($0) }
        var envelope = Data()
        envelope.append(nonceData)
        envelope.append(sealed.ciphertext)
        envelope.append(sealed.tag)
        return envelope.base64EncodedString()
    }

    static func decryptEnvelope(key: SymmetricKey, envelopeBase64: String) throws -> String {
        guard let envelope = Data(base64Encoded: envelopeBase64) else {
            throw SyncError.cryptoFailure
        }
        guard envelope.count >= nonceBytes + tagBytes else {
            throw SyncError.cryptoFailure
        }
        let nonceData = envelope.prefix(nonceBytes)
        let tag = envelope.suffix(tagBytes)
        let ciphertext = envelope.dropFirst(nonceBytes).dropLast(tagBytes)

        let nonce = try AES.GCM.Nonce(data: nonceData)
        let sealed = try AES.GCM.SealedBox(nonce: nonce, ciphertext: ciphertext, tag: tag)
        let plaintext = try AES.GCM.open(sealed, using: key)
        return String(data: plaintext, encoding: .utf8) ?? ""
    }

    enum SyncError: Error {
        case missingPairingCode
        case cryptoFailure
    }
}

// MARK: - PBKDF2 (CommonCrypto)

import CommonCrypto

private func pbkdf2HmacSha256(password: Data, salt: Data, iterations: Int, keyLength: Int) -> Data {
    var derived = Data(count: keyLength)
    let result = derived.withUnsafeMutableBytes { derivedBytes -> Int32 in
        password.withUnsafeBytes { passwordBytes -> Int32 in
            salt.withUnsafeBytes { saltBytes -> Int32 in
                CCKeyDerivationPBKDF(
                    CCPBKDFAlgorithm(kCCPBKDF2),
                    passwordBytes.bindMemory(to: Int8.self).baseAddress,
                    password.count,
                    saltBytes.bindMemory(to: UInt8.self).baseAddress,
                    salt.count,
                    CCPseudoRandomAlgorithm(kCCPRFHmacAlgSHA256),
                    UInt32(iterations),
                    derivedBytes.bindMemory(to: UInt8.self).baseAddress,
                    keyLength
                )
            }
        }
    }
    if result != kCCSuccess {
        return Data(count: keyLength) // zero key on failure; caller will fail to decrypt
    }
    return derived
}

// MARK: - Wire envelope

/// Wire-compatible with `SyncEnvelope` on Windows. Field names use camelCase in JSON.
struct SyncEnvelopeWire: Codable {
    var fingerprint: String
    var kind: String
    var textContent: String?
    var fileName: String?
    var fileUrl: String?
    var payloadDataBase64: String?
    var utiIdentifier: String?
    var createdAt: Date
    var updatedAt: Date
    var isPinned: Bool
    var sourceApplication: String?
    var fromDeviceId: String
    var fromDeviceName: String
}

File: windows/Clipmon/Services/SyncProtocol.cs
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;
using Clipmon.Models;

namespace Clipmon.Services;

/// <summary>
/// Constants and helpers shared by the relay protocol on both ends.
/// IMPORTANT: keep in sync with the macOS Swift implementation.
/// </summary>
public static class SyncProtocol
{
    public const string KeySalt = "clipmon-key-v1";
    public const string RoomSalt = "clipmon-room-v1";
    public const int Pbkdf2Iterations = 200_000;
    public const int KeySizeBytes = 32; // AES-256
    public const int NonceBytes = 12;
    public const int TagBytes = 16;

    public static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
        DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull
    };

    public static byte[] DeriveKey(string pairingCode)
    {
        if (string.IsNullOrEmpty(pairingCode)) throw new ArgumentException("Pairing code is required");
        var salt = Encoding.UTF8.GetBytes(KeySalt);
        using var pbk = new Rfc2898DeriveBytes(pairingCode, salt, Pbkdf2Iterations, HashAlgorithmName.SHA256);
        return pbk.GetBytes(KeySizeBytes);
    }

    public static string DeriveRoomId(string pairingCode)
    {
        if (string.IsNullOrEmpty(pairingCode)) throw new ArgumentException("Pairing code is required");
        var bytes = Encoding.UTF8.GetBytes(pairingCode + ":" + RoomSalt);
        var hash = SHA256.HashData(bytes);
        return Convert.ToHexString(hash).ToLowerInvariant();
    }

    public static string EncryptEnvelope(byte[] key, string plaintextJson)
    {
        var nonce = RandomNumberGenerator.GetBytes(NonceBytes);
        var plaintext = Encoding.UTF8.GetBytes(plaintextJson);
        var ciphertext = new byte[plaintext.Length];
        var tag = new byte[TagBytes];

        using var aes = new AesGcm(key, TagBytes);
        aes.Encrypt(nonce, plaintext, ciphertext, tag);

        var output = new byte[NonceBytes + ciphertext.Length + TagBytes];
        Buffer.BlockCopy(nonce, 0, output, 0, NonceBytes);
        Buffer.BlockCopy(ciphertext, 0, output, NonceBytes, ciphertext.Length);
        Buffer.BlockCopy(tag, 0, output, NonceBytes + ciphertext.Length, TagBytes);
        return Convert.ToBase64String(output);
    }

    /// <summary>
    /// Wire format for Kind matches the Swift raw values exactly: lowercase-first-letter.
    /// Mac uses "text"/"richText"/"spreadsheet"/etc., so Windows MUST send the same strings.
    /// </summary>
    public static string ToWireKind(ClipboardContentKind kind) => kind switch
    {
        ClipboardContentKind.Text => "text",
        ClipboardContentKind.Markdown => "markdown",
        ClipboardContentKind.RichText => "richText",
        ClipboardContentKind.Spreadsheet => "spreadsheet",
        ClipboardContentKind.Image => "image",
        ClipboardContentKind.Audio => "audio",
        ClipboardContentKind.File => "file",
        _ => "text"
    };

    public static ClipboardContentKind FromWireKind(string? wire) => (wire ?? "text").ToLowerInvariant() switch
    {
        "text" => ClipboardContentKind.Text,
        "markdown" => ClipboardContentKind.Markdown,
        "richtext" => ClipboardContentKind.RichText,
        "spreadsheet" => ClipboardContentKind.Spreadsheet,
        "image" => ClipboardContentKind.Image,
        "audio" => ClipboardContentKind.Audio,
        "file" => ClipboardContentKind.File,
        _ => ClipboardContentKind.Text
    };

    public static string DecryptEnvelope(byte[] key, string envelopeBase64)
    {
        var envelope = Convert.FromBase64String(envelopeBase64);
        if (envelope.Length < NonceBytes + TagBytes)
        {
            throw new CryptographicException("Envelope is too short");
        }

        var nonce = envelope.AsSpan(0, NonceBytes);
        var tag = envelope.AsSpan(envelope.Length - TagBytes, TagBytes);
        var ciphertext = envelope.AsSpan(NonceBytes, envelope.Length - NonceBytes - TagBytes);
        var plaintext = new byte[ciphertext.Length];

        using var aes = new AesGcm(key, TagBytes);
        aes.Decrypt(nonce, ciphertext, tag, plaintext);
        return Encoding.UTF8.GetString(plaintext);
    }
}

/// <summary>
/// JSON shape carried inside the encrypted envelope. Wire-compatible with the
/// macOS client. Field names are camelCase in JSON.
/// </summary>
public sealed class SyncEnvelope
{
    public string Fingerprint { get; set; } = string.Empty;
    public string Kind { get; set; } = "Text";
    public string? TextContent { get; set; }
    public string? FileName { get; set; }
    public string? FileUrl { get; set; }
    public string? PayloadDataBase64 { get; set; }
    public string? UtiIdentifier { get; set; }
    public DateTime CreatedAt { get; set; }
    public DateTime UpdatedAt { get; set; }
    public bool IsPinned { get; set; }
    public string? SourceApplication { get; set; }
    public string FromDeviceId { get; set; } = string.Empty;
    public string FromDeviceName { get; set; } = string.Empty;
}
